import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java

```diff
@@ -124,54 +155,87 @@ private void validateParam(String value, String name) throws DicomServiceExcepti
-      IDicomWebClient dicomWebClient) throws Throwable {
-    ExecutorCompletionService<Void> ecs = new ExecutorCompletionService<>(underlyingExecutor);
+  private IDicomWebClient selectDestinationClient(String callingAet, Attributes attrs){
+    for(DestinationFilter filter: destinationMap.keySet()){
+      if(filter.matches(callingAet, attrs)){
+        return destinationMap.get(filter);
+      }
+    }
+    return defaultDicomWebClient;
+  }
 
-    PipedOutputStream pdvPipeOut = new PipedOutputStream();
-    PipedInputStream pdvPipeIn = new PipedInputStream(pdvPipeOut);
-    PipedOutputStream redactedPipeOut = new PipedOutputStream();
-    PipedInputStream redactedPipeIn = new PipedInputStream(redactedPipeOut);
+  private void processStream(Executor underlyingExecutor, InputStream inputStream,
+      List<StreamProcessor> processorList) throws Throwable {
+    if (processorList.size() == 1) {
+      StreamProcessor singleProcessor = processorList.get(0);
+      singleProcessor.process(inputStream, null);
+    } else if (processorList.size() > 1) {
+      List<StreamCallable> callableList = new ArrayList<>();
+
+      PipedOutputStream pdvPipeOut = new PipedOutputStream();
+      InputStream nextInputStream = new PipedInputStream(pdvPipeOut);
+      for(int i=0; i < processorList.size(); i++){
+        StreamProcessor processor = processorList.get(i);
+        InputStream processorInput = nextInputStream;
+        OutputStream processorOutput = null;
+
+        if(i < processorList.size() - 1) {
+          PipedOutputStream pipeOut = new PipedOutputStream();
+          processorOutput = pipeOut;
+          nextInputStream = new PipedInputStream(pipeOut);
+        }
 
-    ecs.submit(() -> {
-      try (redactedPipeIn) {
-        dicomWebClient.stowRs(redactedPipeIn);
+        callableList.add(new StreamCallable(processorInput, processorOutput, processor));
       }
-      return null;
-    });
 
-    ecs.submit(() -> {
-      try (pdvPipeIn) {
-        try (redactedPipeOut) {
-          redactor.redact(pdvPipeIn, redactedPipeOut);
-        }
+      ExecutorCompletionService<Void> ecs = new ExecutorCompletionService<>(underlyingExecutor);
+      for(StreamCallable callable : callableList){
+        ecs.submit(callable);
       }
-      return null;
-    });
-
-    try (pdvPipeOut) {
-      // PDVInputStream is thread-locked
-      StreamUtils.copy(inputStream, pdvPipeOut);
-    } catch (IOException e) {
-      // causes or is caused by exception in redactor.redact, no need to throw this up
-      log.trace("Error copying inputStream to pdvPipeOut", e);
-    }
 
-    try {
-      for (int i = 0; i < 2; ++i) {
-        ecs.take().get();
+      try (pdvPipeOut) {
+        // PDVInputStream is thread-locked
+        StreamUtils.copy(inputStream, pdvPipeOut);
+      } catch (IOException e) {
+        // causes or is caused by exception in redactor.redact, no need to throw this up
```

Comment:
comment seems out of date

Reply:
Changed exact wording, but idea is the same.